File: software/examples/c/example_callback.c

#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_color.h"

#define HOST "localhost"
#define PORT 4223
#define UID "abc" // Change to your UID

// Callback for color with RGBC values greater than 255
void cb_color(uint16_t r, uint16_t g, uint16_t b, uint16_t c, void *user_data) {
	(void)user_data; // avoid unused parameter warning

	printf("Color(R) = %d.\n", r);
    printf("Color(G) = %d.\n", g);
    printf("Color(B) = %d.\n", b);
    printf("Color(C) = %d.\n", c);
	printf("\n");
}

int main() {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	Color c;
	color_create(&c, UID, &ipcon); 

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		exit(1);
	}
	// Don't use device before ipcon is connected

	// Set Period for color callback to 1s (1000ms)
	// Note: The callback is only called every second if the 
	//       color has changed since the last call!
	color_set_color_callback_period(&c, 1000);

	// Register color callback to function cb_color
	color_register_callback(&c,
	                              COLOR_CALLBACK_COLOR, 
	                              (void *)cb_color,
	                              NULL);

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
}

File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/color_bricklet"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Color Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	c, _ := color_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	c.RegisterColorCallback(func(r uint16, g uint16, b uint16, c uint16) {
		fmt.Printf("Color [R]: %d\n", r)
		fmt.Printf("Color [G]: %d\n", g)
		fmt.Printf("Color [B]: %d\n", b)
		fmt.Printf("Color [C]: %d\n", c)
		fmt.Println()
	})

	// Set period for color receiver to 1s (1000ms).
	// Note: The color callback is only called every second
	//       if the color has changed since the last call!
	c.SetColorCallbackPeriod(1000)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
